src/Mvc/Mvc.RazorPages/src/Infrastructure/PageLoaderMatcherPolicy.cs

```diff
@@ -73,17 +72,44 @@ public Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context,
             for (var i = 0; i < candidates.Count; i++)
             {
                 ref var candidate = ref candidates[i];
-                var endpoint = (RouteEndpoint)candidate.Endpoint;
+                var endpoint = candidate.Endpoint;
 
                 var page = endpoint.Metadata.GetMetadata<PageActionDescriptor>();
                 if (page != null)
                 {
-                    var compiled = _loader.Load(page);
```

Comment:
This gets called on every request and we produce a new instance of `CompiledPageActionDescriptor`. In the past the ADP cached this for us, which now has to be pushed to the loader.